File: src/ObjectManager/Object.Tes/FilePacks/Records/340-PGRD.Path grid.cs
using OA.Core;
using System;
using System.Collections.Generic;
using UnityEngine;

namespace OA.Tes.FilePacks.Records
{
    public class PGRDRecord : Record
    {
        public struct DATAField
        {
            public int X;
            public int Y;
            public short Granularity;
            public short PointCount;

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId formatId)
            {
                if (formatId == GameFormatId.TES3)
                {
                    X = r.ReadLEInt32();
                    Y = r.ReadLEInt32();
                    Granularity = r.ReadLEInt16();
                    PointCount = r.ReadLEInt16();
                    return;
                }
                else X = Y = Granularity = 0;
                PointCount = r.ReadLEInt16();
            }
        }

        public struct PGRPField
        {
            public Vector3 Point;
            public byte Connections;

            public PGRPField(UnityBinaryReader r, int dataSize)
            {
                Point = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
                Connections = r.ReadByte();
                r.SkipBytes(3); // Unused
            }
        }

        public struct PGRRField
        {
            public short StartPointId;
            public short EndPointId;

            public PGRRField(UnityBinaryReader r, int dataSize)
            {
                StartPointId = r.ReadLEInt16();
                EndPointId = r.ReadLEInt16();
            }
        }

        public struct PGRIField
        {
            public short PointId;
            public Vector3 ForeignNode;

            public PGRIField(UnityBinaryReader r, int dataSize)
            {
                PointId = r.ReadLEInt16();
                r.SkipBytes(2); // Unused (can merge back)
                ForeignNode = new Vector3(r.ReadLESingle(), r.ReadLESingle(), r.ReadLESingle());
            }
        }

        public struct PGRLField
        {
            public FormId<REFRRecord> Reference;
            public short[] PointIds;

            public PGRLField(UnityBinaryReader r, int dataSize)
            {
                Reference = new FormId<REFRRecord>(r.ReadLEUInt32());
                PointIds = new short[(dataSize - 4) >> 2];
                for (var i = 0; i < PointIds.Length; i++)
                {
                    PointIds[i] = r.ReadLEInt16();
                    r.SkipBytes(2); // Unused (can merge back)
                }
            }
        }

        public override string ToString() => $"PGRD: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public DATAField DATA; // Number of nodes
        public PGRPField[] PGRPs;
        public UNKNField PGRC;
        public UNKNField PGAG;
        public PGRRField[] PGRRs; // Point-to-Point Connections
        public List<PGRLField> PGRLs; // Point-to-Reference Mappings
        public PGRIField[] PGRIs; // Inter-Cell Connections

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize, formatId); return true;
                case "PGRP": PGRPs = new PGRPField[dataSize >> 4]; for (var i = 0; i < PGRPs.Length; i++) PGRPs[i] = new PGRPField(r, 16); return true;
                case "PGRC": PGRC = new UNKNField(r, dataSize); return true;
                case "PGAG": PGAG = new UNKNField(r, dataSize); return true;
                case "PGRR": PGRRs = new PGRRField[dataSize >> 2]; for (var i = 0; i < PGRRs.Length; i++) PGRRs[i] = new PGRRField(r, 4); r.SkipBytes(dataSize % 4); return true;
                case "PGRL": if (PGRLs == null) PGRLs = new List<PGRLField>(); PGRLs.Add(new PGRLField(r, dataSize)); return true;
                case "PGRI": PGRIs = new PGRIField[dataSize >> 4]; for (var i = 0; i < PGRIs.Length; i++) PGRIs[i] = new PGRIField(r, 16); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/340-PGRD.Path grid.swift
//
//  PGRDRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//
import SceneKit

public class PGRDRecord: Record {
    public struct DATAField {
        public let x: Int32
        public let y: Int32
        public let granularity: Int16
        public let pointCount: Int16

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format == .TES3 else {
                x = 0; y = 0; granularity = 0
                pointCount = r.readLEInt16()
                return
            }
            x = r.readLEInt32()
            y = r.readLEInt32()
            granularity = r.readLEInt16()
            pointCount = r.readLEInt16()
        }
    }

    public typealias PGRPField = (
        point: Float3,
        connections: UInt8,
        pad01: UInt8,
        pad02: UInt16
    )

    public typealias PGRRField = (
        startPointId: Int16,
        endPointId: Int16
    )

    public typealias PGRIField = (
        pointId: Int16,
        pad02: UInt16,
        foreignNode: Float3
    )

    public struct PGRLField {
        public let reference: FormId<REFRRecord>
        public var pointIds: [Int16]

        init(_ r: BinaryReader, _ dataSize: Int) {
            reference = FormId<REFRRecord>(r.readLEUInt32())
            pointIds = [Int16](); let capacity = (dataSize - 4) >> 2; pointIds.reserveCapacity(capacity)
            for _ in 0..<capacity {
                pointIds.append(r.readLEInt16())
                r.skipBytes(2) // Unused (can merge back)
            }
        }
    }

    public override var description: String { return "PGRD: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var DATA: DATAField! // Number of nodes
    public var PGRPs: [PGRPField]!
    public var PGRC: UNKNField!
    public var PGAG: UNKNField!
    public var PGRRs: [PGRRField]! // Point-to-Point Connections
    public var PGRLs: [PGRLField]? = nil// Point-to-Reference Mappings
    public var PGRIs: [PGRIField]! // Inter-Cell Connections

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "DATA": DATA = DATAField(r, dataSize, format)
        case "PGRP": PGRPs = r.readTArray(dataSize, count: dataSize >> 4)
        case "PGRC": PGRC = r.readBYTV(dataSize)
        case "PGAG": PGAG = r.readBYTV(dataSize)
        case "PGRR": PGRRs = r.readTArray(dataSize, count: dataSize >> 2) //r.skipBytes(dataSize % 4)
        case "PGRL": if PGRLs == nil { PGRLs = [PGRLField]() }; PGRLs!.append(PGRLField(r, dataSize))
        case "PGRI": PGRIs = r.readTArray(dataSize, count: dataSize >> 4)
        default: return false
        }
        return true
    }
}
